### PR TITLE
feat(agents-api): Add static task validations to catch errors beforehand

### LINKED ISSUE #535: feat: Static checks of python expressions inside workflows
Consider using [pysa](https://pyre-check.org/docs/pysa-advanced/)

### PR DESCRIPTION
**Pull request by creatorrr**
### **User description**
Signed-off-by: Diwank Singh Tomer <diwank.singh@gmail.com>

Closes #535

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced static validation for Python expressions in workflows.
  - Added `validate_py_expression` to statically validate Python expressions.
  - Implemented `backwards_compatibility` for expression transformation.

- Enhanced task validation with detailed error reporting.
  - Integrated validation into task creation and update endpoints.
  - Added structured error responses for invalid task specifications.

- Refactored and improved API call execution with better error handling.
  - Enhanced logging and error reporting for HTTP requests.
  - Added support for base64 encoding of response content.

- Added comprehensive test coverage for new validation features.
  - Included tests for expression validation, task validation, and edge cases.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>execute_api_call.py</strong><dd><code>Refactored API call execution with improved error handling</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-bbd41da2245c7eb5f9888933d9205d067d35df6496f6fc3581fb28619b7a1b52">+63/-20</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>base_evaluate.py</strong><dd><code>Enhanced expression evaluation with backwards compatibility</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-05b346521c23f568c9804ac59931eb4fd5a61ec88110a783debb1ed329b47470">+18/-26</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>tasks.py</strong><dd><code>Added input limit parameter to task preparation methods</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-8a1abcedcd95d0217595a53a9c503fa4dec969579ce9cf9a120193413758fbef">+6/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>task_validation.py</strong><dd><code>Added static validation for Python expressions and task specifications</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-536283ced20f35bc610e4d49258098f86053911c0e53af09a035b6dffe9fcadc">+531/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>create_or_update_task.py</strong><dd><code>Integrated task validation into update endpoint</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-cb62c71de6ceaafaae0f7f0935a77927b03bf8d1967bc91fadf6164078f200eb">+49/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>create_task.py</strong><dd><code>Integrated task validation into creation endpoint</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-60c05ef9fdd03b8f6ad6639b8f2d2ef1e8ec2041662c06c50134bec6d35ae8db">+49/-7</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>test_base_evaluate.py</strong><dd><code>Added tests for expression evaluation and backwards compatibility</code></dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-6903501fdd4934d40307d1a19f64e8894d7b68754833ef9c6543c67cb0b66430">+77/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_expression_validation.py</strong><dd><code>Added tests for Python expression validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-427f8389d4ec97bdd68a96bd1850d3bf3f670a1927ee01c9728295bd78701ca9">+173/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_task_validation.py</strong><dd><code>Added tests for task validation and nested workflows</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/julep-ai/julep/pull/1263/files#diff-3fd2df298d814764cdf747caab177be03452c45b550d1edbf8e3e6e847eb9253">+316/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

----

> [!IMPORTANT]
> Introduces static validation for Python expressions in workflows, enhances task validation, refactors API call execution, and adds comprehensive tests.
> 
>   - **Static Validation**:
>     - Added `validate_py_expression` for static validation of Python expressions in `task_validation.py`.
>     - Implemented `backwards_compatibility` for expression transformation.
>   - **Task Validation**:
>     - Integrated validation into task creation (`create_task.py`) and update (`create_or_update_task.py`) endpoints.
>     - Added structured error responses for invalid task specifications.
>   - **API Call Execution**:
>     - Refactored `execute_api_call()` in `execute_api_call.py` with improved error handling and logging.
>     - Added base64 encoding for response content.
>   - **Testing**:
>     - Added tests for expression validation in `test_expression_validation.py` and `test_task_validation.py`.
>     - Included tests for `base_evaluate()` in `test_base_evaluate.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for 55120ee78e23a73f8a7f3179ccc5edbcef523bf1. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/agents_api/common/utils/task_validation.py [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+
+    # Check for undefined names
+    referenced_names = {
+        node.id
+        for node in name_nodes
+        if isinstance(node, ast.Name) and isinstance(node.ctx, ast.Load)
+    }
+
+    # Import ALLOWED_FUNCTIONS and stdlib from where they're defined
+    from ...activities.utils import ALLOWED_FUNCTIONS, stdlib
```
Comment (whiterabbit1983) [inline agents-api/agents_api/common/utils/task_validation.py:139]: can this be imported at the beginning of the file?

Comment (creatorrr) [reply agents-api/agents_api/common/utils/task_validation.py:139]: causes a circular import :-/

File: agents-api/agents_api/common/protocol/tasks.py [hunk 2]
```diff
@@ -236,7 +236,9 @@ def model_dump(self, *args, **kwargs) -> dict[str, Any]:
 
         return dump | execution_input
 
-    async def get_inputs(self) -> tuple[list[Any], list[str | None], dict[str, Any]]:
+    async def get_inputs(
+        self, limit: int = 50
```
[the remark this answers is not in the document]
Comment (Ahmad-mtos) [reply agents-api/agents_api/common/protocol/tasks.py:240]: not sure about making it configurable, but at least I know this limit would cause issues for not so extreme edge cases. I suggest raising the limit for the time being until we optimize the list execution transition query

File: agents-api/agents_api/common/utils/task_validation.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+        return issues
+
+    # Apply backwards compatibility transformation first
+    expr = backwards_compatibility(expr)
+
+    # Ensure the expression is stripped before checking prefix
+    expr = expr.strip()
+
+    # Handle expressions with $ prefix - return early if it doesn't start with $
+    if not expr.startswith("$"):
```
Comment (Ahmad-mtos) [inline agents-api/agents_api/common/utils/task_validation.py:97]: why do we need the first check? isn't the second one sufficient? ("$', "$ ")

File: agents-api/agents_api/common/utils/task_validation.py [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+
+    # The space after $ is required for expressions to be evaluated as Python code
+    if not expr.startswith("$ "):
+        return issues
+
+    # Remove $ and strip any leading space after $
+    expr = expr[1:].strip()
+
+    # Special case: just a $ sign with nothing after it
+    if not expr:
```
Comment (Ahmad-mtos) [inline agents-api/agents_api/common/utils/task_validation.py:108]: not possible, since we've already stripped expr and checked for (startswith "$ ")

File: agents-api/agents_api/common/utils/task_validation.py [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+        return issues
+
+    # Handle f-string expressions (these are often used in templates)
+    expr = expr.strip()
```
Comment (Ahmad-mtos) [inline agents-api/agents_api/common/utils/task_validation.py:112]: redundant strip

File: agents-api/agents_api/common/utils/task_validation.py [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+        set(ALLOWED_FUNCTIONS.keys())
+        | set(stdlib.keys())
+        | {"true", "false", "null", "NEWLINE", "hasattr"}
+    )
+
+    # Add standard placeholder variable names if allowed
+    if allow_placeholder_variables:
+        allowed_names.add("_")  # Special underscore variable
+        # Some common inputs that might be accessed via _
+        allowed_names.update({"inputs", "outputs", "state"})
```
Comment (Ahmad-mtos) [inline agents-api/agents_api/common/utils/task_validation.py:152]: what about new syntax? (steps[x].input...)

Comment (Ahmad-mtos) [reply agents-api/agents_api/common/utils/task_validation.py:152]: I just checked the tests, I didn't see a test covering this case

File: agents-api/agents_api/common/protocol/tasks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (creatorrr) [reply agents-api/agents_api/common/protocol/tasks.py:241]: made configurable, and set default to 250
